native-maven-plugin/build.gradle.kts

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2020, 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * The Universal Permissive License (UPL), Version 1.0
+ *
+ * Subject to the condition set forth below, permission is hereby granted to any
+ * person obtaining a copy of this software, associated documentation and/or
+ * data (collectively the "Software"), free of charge and under any and all
+ * copyright rights in the Software, and any and all patent rights owned or
+ * freely licensable by each licensor hereunder covering either (i) the
+ * unmodified Software as contributed to or provided by such licensor, or (ii)
+ * the Larger Works (as defined below), to deal in both
+ *
+ * (a) the Software, and
+ *
+ * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+ * one is included with the Software each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ *
+ * The above copyright notice and either this complete permission notice or at a
+ * minimum a reference to the UPL must be included in all copies or substantial
+ * portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+import com.bmuschko.gradle.docker.tasks.image.DockerBuildImage
+import com.bmuschko.gradle.docker.tasks.image.Dockerfile
+import org.graalvm.build.maven.GeneratePluginDescriptor
+
+plugins {
+    `java-library`
+    groovy
+    checkstyle
+    `java-test-fixtures`
+    id("org.graalvm.build.java")
+    id("org.graalvm.build.publishing")
+    id("org.graalvm.build.maven-functional-testing")
+    id("com.bmuschko.docker-remote-api") version "7.1.0"
```

Comment:
I believe including docker here is a overkill. Our CI already has Maven and Gradle preinstalled, and we include a script for downloading and configuring GraalVM and Native Image (that should be replaced once we start providing toolchain fetching).

If nothing, debug log is filled with Docker logs and dependency resolutions (first image build starts at line 13k).
Also this is adding a new dependency (with potential compliance issues).

If fetching maven is the issue here, we can create a `mvnw` which is now upstreamed and official way to fetch maven.  
If you believe that we should Dockerize our builds, I'm for that, but I believe that it should be at a repository level, and that we should run root Gradle project inside of said Docker environment.

Reply:
Did you look at the [commit message introducing this](https://github.com/graalvm/native-build-tools/pull/87/commits/d96237b8242f1c1054c76ec4ba925590c873e940) ? The problem is not downloading Maven, it's having reproducible builds. With Maven it's barely impossible to have reliable builds because it insists on copying everything into the `.m2` repository. This means that:

- builds are order dependent: depending on what order you execute the dependent projects, dependencies might be cached, including snapshot dependencies, ignoring local changes unless you run systematically with `-U`
- we cannot cache the external dependencies from locally built dependencies in isolation, meaning that as soon as we build one project (say the junit native extension), dependencies are put in .m2 _but also_ the project themselves, making it impossible to have integration tests which start from a clean state _without having to redownload everything everytime_

So what this does is basically introducing docker in order to be able to properly isolate test executions, and not having tests accidentally passing because another test passed. To reduce build time, we create a container which _predownloads_ the external dependencies we need, and each test uses that image. While it introduces a layer of complexity, this is *much more reliable* than other mechanisms: any other way would require a way to clean up the local .m2 repository from "pollution" of the tests execution (capture state before and after).

As for the compliance thing, again I can't imagine how this could be an issue.

Reply:
I don’t see how this impacts compliance it is just a build time dependency and ensures we have better Maven tests

Reply:
I still believe that all of that could be achieved without the dependency on Docker. 
For example you can:
* use the `mvnw` to download separate distribution of maven
* [change the `.m2` location ](https://stackoverflow.com/a/57223696) to the empty directory
* run the bootstrapping project and inject the fake repository
* save copy of the resulting repository in order to re-inject it after subsequent test execution

Current state seems to me as if we are shooting a fly with a cannon. However, I'm fine with merging this as is and fixing it in a follow-up PR in order to move faster. One thing that should be changed prior to merging this is the removal of [GraalVM downloading step in the CI](https://github.com/melix/native-build-tools/blob/051df2f64bf0b0a29f012ba032a85d518a763fb3/.github/workflows/native-maven-plugin.yml#L26-L31) since we are already downloading GraalVM through a Docker image.

Reply:
This is still significantly slower, in particular for local builds, since for every change you'd have to re-download every single dependency Maven needs. The docker images can be layered which makes their build much faster, especially for local builds (for CI it doesn't matter yet since we don't use caching, but we could).

Also even if you change the .m2 location, you'd have to use a _separate_ m2 location for every single test. And your local settings.xml (in your user home) would influence the build. It's barely impossible to properly isolate Maven builds, unfortunately.

Reply:
> One thing that should be changed prior to merging this is the removal of GraalVM downloading step in the CI since we are already downloading GraalVM through a Docker image.

Good catch!